instrumentation/rspec/lib/opentelemetry/instrumentation/rspec/formatter.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'opentelemetry-api'
+require_relative './version'
+
+module OpenTelemetry
+  module Instrumentation
+    module RSpec
+      # An RSpec Formatter that outputs Otel spans
+      class Formatter
+        attr_reader :output
+
+        ::RSpec::Core::Formatters.register self, :example_started, :example_finished, :example_group_started, :example_group_finished, :start, :stop
+
+        @clock = Time.method(:now)
+
+        def self.current_timestamp
+          @clock.call
+        end
+
+        def initialize(output = StringIO.new, tracer_provider = OpenTelemetry.tracer_provider)
+          @spans_and_tokens = []
+          @output = output
+          @tracer_provider = tracer_provider
+        end
+
+        def tracer
+          @tracer ||= @tracer_provider.tracer('OpenTelemetry::Instrumentation::RSpec', OpenTelemetry::Instrumentation::RSpec::VERSION)
+        end
+
+        def current_timestamp
+          self.class.current_timestamp
+        end
+
+        def start(notification)
+          span = tracer.start_span('RSpec suite', start_timestamp: current_timestamp)
+          track_span(span)
+        end
+
+        def stop(notification)
+          pop_and_finalize_span
+        end
+
+        def example_group_started(notification)
+          description = notification.group.description
+          span = tracer.start_span(description, start_timestamp: current_timestamp)
+          track_span(span)
+        end
+
+        def example_group_finished(notification)
+          pop_and_finalize_span
+        end
+
+        def example_started(notification)
+          example = notification.example
+          attributes = {
+            'location' => example.location.to_s,
+            'full_description' => example.full_description.to_s,
+            'described_class' => example.metadata[:described_class].to_s
+          }
+          span = tracer.start_span(example.description, attributes: attributes, start_timestamp: current_timestamp)
+          track_span(span)
+        end
+
+        def example_finished(notification)
+          pop_and_finalize_span do |span|
+            result = notification.example.execution_result
+            notification.example.metadata
+
+            span.set_attribute('result', result.status.to_s)
+
+            if (exception = result.exception)
+              span.record_exception(exception)
+              span.set_attribute('message', exception.message) if exception.is_a? ::RSpec::Expectations::ExpectationNotMetError
```

Comment:
Initially, it was to try and show a distinction between examples that error and examples that fail, but in practice it just adds duplication so I’m happy to remove.

Reply:
I had a look at this. My current preference is to keep the failure message attached to the span, as `rspec.example.failure_message` rather than on an attached span event so that it can be found easily next to the `result` attribute. A lot of this is dependent on how tracing tools visualize the trace, span, span events.

It would also be convenient to be able to do queries of the span data like:
```
select description, failure_message from spans where result = false
```

Reply:
Actually, would adding this as the description to the `span.status` be sensible?

Reply:
hm, yes I think this would fit nicely into the status description if possible.

Thinking a bit more, I am fine with leaving it attached in a few places tbh, not the end of the world, we'd just probably want to clean it up before moving it out of experimental.  

I recognize in practice that basically all observability UIs out there do a poor job of showing span events, but imo it's a bit of their responsibility to support OTEL better, less than it is our responsibility to ensure compatibility. That being said i don't want having this live in -contrib degrade your experience using it as a user, so fine with this for now.

Reply:
OK, I've added it as the description now.